R/trs_low_high_thresholds_for_ecoinvent_all.R

```diff
@@ -0,0 +1,100 @@
+#' @param scenarios A dataframe
+#'
+#' @keywords internal
+#'
+#' @return A dataframe.
+#' @export
+#'
+#' @examples
+#' library(tiltToyData)
+#' library(readr)
+#' library(dplyr)
+#' options(readr.show_col_types = FALSE)
+#'
+#' emissions_profile_products <- read_csv(toy_emissions_profile_products_ecoinvent())
+#' all_uuids_scenario_sectors <- read_csv(toy_sector_profile_companies()) |>
+#'   select(-c("companies_id", "company_name", "clustered")) |>
+#'   distinct()
+#' scenarios <- read_csv(toy_sector_profile_any_scenarios())
+#'
+#' output <- trs_low_high_thresholds_for_ecoinvent_all(
+#'   emissions_profile_products,
+#'   all_uuids_scenario_sectors,
+#'   scenarios
+#' )
+#' output
+trs_low_high_thresholds_for_ecoinvent_all <- function(emissions_profile_products,
+                                                      all_uuids_scenario_sectors,
+                                                      scenarios) {
+  epa_profile_ranking <- epa_compute_profile_ranking(emissions_profile_products) |>
+    prepare_profile_ranking()
+
+  spa_reduction_targets <- spa_compute_profile_ranking(
+    all_uuids_scenario_sectors,
+    scenarios
+  ) |>
+    prepare_reduction_targets()
+
+  dplyr::full_join(
+    epa_profile_ranking,
+    spa_reduction_targets,
+    by = c("activity_uuid_product_uuid"),
+    relationship = "many-to-many"
+  ) |>
+    create_tr_benchmarks_tr_score(.data$profile_ranking, .data$reductions) |>
+    distinct() |>
+    add_trs_thresholds(.by = "benchmark_tr_score")
+}
+
+#' Calulate `transition_risk_score` and `benchmark_tr_score` columns
+#'
+#' @param data Dataframe.
+#' @param profile_ranking Dataframe column.
+#' @param reduction_targets Dataframe column.
+#' @keywords internal
+#' @export
+create_tr_benchmarks_tr_score <- function(data, profile_ranking, reduction_targets) {
+  mutate(
+    data,
+    transition_risk_score = ifelse(
+      is.na({{ profile_ranking }}) | is.na({{ reduction_targets }}),
+      NA,
+      ({{ profile_ranking }} + {{ reduction_targets }}) / 2
+    ),
+    benchmark_tr_score = ifelse(
+      is.na({{ profile_ranking }}) | is.na({{ reduction_targets }}),
+      NA,
+      paste(.data$scenario_year, .data$grouped_by, sep = "_")
+    )
+  )
+}
+
+add_trs_thresholds <- function(data, .by) {
+  mutate(data,
+    trs_low_threshold = stats::quantile(.data$transition_risk_score,
+      probs = c(1 / 3, 2 / 3),
+      na.rm = TRUE
+    )[[1]],
+    trs_high_threshold = stats::quantile(.data$transition_risk_score,
+      probs = c(1 / 3, 2 / 3),
+      na.rm = TRUE
+    )[[2]],
+    .by = all_of(.by)
+  )
+}
+
+prepare_reduction_targets <- function(data) {
+  data |>
+    select(c("activity_uuid_product_uuid", "scenario", "year", "reductions")) |>
+    mutate(scenario_year = paste(.data$scenario, .data$year, sep = "_")) |>
+    select(-c("scenario", "year"))
+}
+
+prepare_profile_ranking <- function(data) {
+  data |>
+    select(c("activity_uuid_product_uuid", "grouped_by", "profile_ranking"))
+}
```

Comment:
Everything here seems to be self-contained. It only depends on `tiltIndicator::*compute_profile_ranking()`. If tiltIndicator exports those two functions, then you can code everything elsewhere. 

Thinking about how our system has evolved, I think the best place is a new package tiltTransitionRisk

```
INDICATORS
--------------------------
tiltIndicator (tiltProfiles?)
* `emissions_profile()`
* `sector_profile()`
tiltTransitionRisk ^1
* `transition_risk()`
==========================>>> tiltIndicatorAfter
                              * `profile_emissions()                              
                              * `profile_sector()                              
                              * `score_transition_risk()`
==========================================================>>> tiltPlot
                                                              * ...
=======================================================================>>> tiltWebTool
                                                                           * https://some/url/
   
                                                                       
^1: This is what we seem to need now: a separate tiltTransitionRisk module
```

https://github.com/2DegreesInvesting/tiltTransitionRisk

Reply:
RE name suggestions see https://github.com/2DegreesInvesting/tiltTransitionRisk/discussions/9

Reply:
@kalashsinghal, also here are my general suggestions, that might help reshape this PR before it lands in tiltTransitionRisk:

- [ ] Rename the PR to reflect that it adds a function to the package, not a big dataset.The function may be used with a big dataset, but it may as well work with tiny datasets in tests.
- [ ] Pass threshods as arguments (to `quantile()`)
- [ ] Spell out names , e.g. `trs` to `transition_risk_score`. Benefits: a) I can more quickly understand and help with the review; b) we can "test" which names seem good (evocative, consistent, and concise). We'll need that lesson to decide the name of the most user-facing interfaces. We can rename to more terse names after final review.
- [ ] Rename `prepare*()` to `polish*()`. Use "prepare" for functions that adapt inputs, and "polish" for functions that adapt outputs.

Reply:
@maurolepore Please note that the low and high thresholds of transition risk scores created in this PR will be used inside the [score_transition_risk](https://github.com/2DegreesInvesting/tiltIndicatorAfter/blob/main/R/score_transition_risk.R) function present in tiltIndicatorAfter. 

If we complete this PR in another package, then we might have to move this function in that package as well. Do you think we should also transfer all the code related to transition risk score in tiltIndicatorAfter to another package?

Reply:
Sure, tiltIndicatorAfter would then import those functions from tiltTransitionRisk and use them.

Reply:
Makes sense to me! @AnneSchoenauer @Tilmon Do you agree?

Reply:
So the code will look like this:

```r
tiltIndicatorAfter::score_transition_risk <- function(...) {
  ...
  tiltTransitionRisk::categorize_extreme_transition_risk(...)
  ...
}
```

Reply:
This idea is expressed in this part of my sketch above. Here you can see that tiltIndicatorAfter depends on both tiltIndicator and also tiltTransitionRisk.

```
tiltIndicator (tiltProfiles?)
* `emissions_profile()`
* `sector_profile()`
tiltTransitionRisk ^1
* `transition_risk()`
==========================>>> tiltIndicatorAfter
                              * `profile_emissions()                              
                              * `profile_sector()                              
                              * `score_transition_risk()`
```

Reply:
RE https://github.com/2DegreesInvesting/tiltIndicator/pull/790#discussion_r1620835984 

@AnneSchoenauer @Tilmon before you struggle to answer this question:

1. Moving  this PR to a different package is easy. The package [alerady exists](https://github.com/2DegreesInvesting/tiltTransitionRisk), and @kalashsinghal nicely developed the code in independent files. So we will just need to move files from one folder to another folder.

2. The value is hard for you to see. The proposed architecture does reflect the conceptual separation of (emissions and sector) from (transition risk). But the reasons a system's architecture evolve are mainly driven by develop-ability. 
The benefits are technical and hard to see from the outside. For example, the benefit of having [an independent module about "CO2" ](https://github.com/2DegreesInvesting/tiltAddCo2) allowed me to run many tests much much faster than I could have done it in tiltIndicatorAfter (which does a lot more, and slower things).

Reply:
For completion, I bring here a conversation I just had with @kalashsinghal 

I hear the preference is to keep this PR in tiltIndicator for now. Sounds good. 

Please keep working on separate, new files so that I can easily move them from tiltIndicator/R and tiltIndicator/tests/ to tiltTransitionRisk/R and tiltTransitionRisk/tests/.


See my suggestions here
https://github.com/2DegreesInvesting/tiltIndicator/pull/790#discussion_r1620823281

Reply:
@maurolepore Thanks for the updating this on the thread! And I will keep on working as you suggested :)